BUG: spatial.HalfspaceIntersection: raise on non-feasible half space

tylerjereddy opened the pull request:
* Fixes #19865

* Rather than returning a problematic half space intersection polytope, we raise an error for incremental processing when a new half space is added for which the interior point is no longer clearly interior

[skip cirrus]

tylerjereddy committed:
BUG: raise on non-feasible half space

* Fixes #19865

* Rather than returning a problematic half space intersection
polytope, we raise an error for incremental processing when
a new half space is added for which the interior point is no
longer clearly interior

[skip cirrus]

Co-authored-by: Richard Strong Bowen <richard@geomagical.com>

tylerjereddy commented on scipy/spatial/_qhull.pyx:
```diff
@@ -2870,6 +2870,15 @@ class HalfspaceIntersection(_QhullUser):
+            dist += np.dot(halfspace[:dims], self.interior_point[:dims])
+            if dist > 0:
+                msg = f"feasible point is not clearly inside halfspace: {halfspace}"
+                raise QhullError(msg)
```
@RichardBowenGM what do you think of this shim? I suppose we could try to vectorize it (for performance) and potentially report the full set of violating half spaces as well, rather than just the first one encountered, though this may be half decent as a start.

Also, I'm happy to add you as a commit co-author if you tell me what to put per: https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors

tylerjereddy commented on scipy/spatial/_qhull.pyx:
```diff
@@ -2870,6 +2870,15 @@ class HalfspaceIntersection(_QhullUser):
+        # We check for non-feasibility of incremental additions
+        # in a manner similar to `qh_sethalfspace`
+        dims = halfspaces.shape[1] - 1
+        for halfspace in halfspaces:
```
Our testsuite may not be sensitive to it, but now that I look at this, I suspect I/we should check for the case where a single violating halfspace is added incrementally rather than a bunch of them. My intuition is we may get the wrong behavior in that case.

tylerjereddy replied:
I just mean that if you feed in a 1d array like `[0.4, 0.4, 0.3]` for the new halfspace that line might iterate over the elements instead of the half spaces in a 2d array as currently tested.

tylerjereddy commented:
I updated the test to include a 1D half space, and fixed the code to support that. I added Richard as a co-author.

I haven't managed to "vectorize" the code though, the suggestion above didn't quite pass the full suite yet.

tylerjereddy commented:
I saw a real `spatial` failure actually, but it looks like you changed the code a bit from `>=` to `>` (which is more consistent with what I did originally, so makes sense).

I did see one other issue--the index in the original error message always seemed to be `1` even if there was a single halfspace involved, so that gave me some pause.

Your other failures don't look related.

tylerjereddy commented:
I probably need a little more time to think about this before branching for `1.13.0` in a few days, so I'll bump the milestone.

ev-br commented:
Where does this PR stand, now that 1.14 is closing in?

tylerjereddy commented:
> Where does this PR stand, now that 1.14 is closing in?

Heh yeah, I don't do much real SWE anymore it seems. I'll try to take a look if I get a chance, but I don't think we'd get too many complaints if we had to delay the fix a bit.

tylerjereddy commented:
I do need to get back to this yes.

tylerjereddy commented:
I need to find time to circle back here, but realistically not likely to happen with branching scheduled in a few days, so I'll bump the milestone one more time.

tylerjereddy committed:
TST, MAINT: PR 20035 revisions

* `test_gh_19865()` was simplified to avoid unnecessary looping
for 2D input; the regex match on the error message now enforces
identification of an appropriate violating half space.

* Also vectorize the feasibility check, and add a 3D test that
simultaneously enforces the closed half spaces.

Co-authored-by: Richard Strong Bowen <richard@geomagical.com>

tylerjereddy commented on scipy/spatial/_qhull.pyx:
```diff
@@ -2929,6 +2929,20 @@ class HalfspaceIntersection(_QhullUser):
+        dists = np.dot(halfspaces[:, :self.ndim], self.interior_point) + halfspaces[:, -1]
+        # HalfspaceIntersection uses closed half spaces so
+        # the feasible point also cannot be directly on the boundary
+        viols = dists >= 0
```
`HalfspaceIntersection` docs suggest that the half spaces it accepts are "closed" so I think we should indeed reject interior points directly on the boundary like this. I believe @RichardBowenGM and I agree on this.

tylerjereddy commented on scipy/spatial/_qhull.pyx:
```diff
@@ -2929,6 +2929,20 @@ class HalfspaceIntersection(_QhullUser):
+        # We check for non-feasibility of incremental additions
+        # in a manner similar to `qh_sethalfspace`
+        halfspaces = np.atleast_2d(halfspaces)
+        dists = np.dot(halfspaces[:, :self.ndim], self.interior_point) + halfspaces[:, -1]
```
This is now vectorized in a manner similar to that described by @RichardBowenGM, and I've added a 3d test as well, just to be cautious.

tylerjereddy commented on scipy/spatial/tests/test_qhull.py:
```diff
@@ -1181,6 +1181,45 @@ def test_cube(self):
+                                                              np.zeros(3),
+                                                              incremental=True)
+        with pytest.raises(qhull.QhullError, match="feasible.*[1 0 0 0]"):
+            incremental_intersector.add_halfspaces(halfspaces)
```
This test simultaneously verifies that 3D behavior is appropriate and that adding a half space that exactly intersects the feasibility point is rejected (enforcing closed half spaces).

tylerjereddy commented on scipy/spatial/tests/test_qhull.py:
```diff
@@ -1181,6 +1181,45 @@ def test_cube(self):
+        # feasibility
+        halfspaces = np.array([[1, 1, 1, -1], # doesn't exclude origin
+                               [-1, -1, -1, -1], # doesn't exclude origin
+                               [1, 0, 0, 0]]) # the origin is on the line
```
at least one thing left to do here--if I remove the bad half space, the test will error out with `ValueError: operands could not be broadcast together with shapes (2,3) (2,)` so we may need to be cautious about `halfspaces = np.atleast_2d(halfspaces)` -- our testsuite isn't very thorough on this obviously

tylerjereddy reviewed (commented):
Added some updates/comments

tylerjereddy commented:
@rsbown when you have time, can you please review this again now that gh-22693 has been merged? We try to avoid merging our own work as much as possible. Don't be shy about asking for improvements to the testing, etc., now that you've thought about the problem a bit more.

rsbowen commented on scipy/spatial/_qhull.pyx:
```diff
@@ -2935,6 +2935,20 @@ class HalfspaceIntersection(_QhullUser):
 
         """
+        # We check for non-feasibility of incremental additions
+        # in a manner similar to `qh_sethalfspace`
```
To be extra thorough, we should check and raise here if `halfspaces.ndim > 2`.

rsbowen reviewed (approved):
LGTM; just a few small comments.

Testing looks good to me -- it never tests the happy path, but other tests of this function do. I would suggest a test for the case where we are adding a single halfspace of shape `(ndim+1,)` rather than of shape `(1, ndim+1)`, to test the `atleast_2d` codepath.

We could be super thorough and parameterize the test over `ndim`, but I think that's overkill.

tylerjereddy committed:
MAINT: PR 20035 revisions

* `add_halfspaces()` now provides a more helpful error
message when receiving an array of halfspaces with more
than two dimensions

* use `viols.any()` to check for violations based on reviewer
feedback

* add a regression test for acceptance of 1D `halfspaces` input
to `add_halfspaces()`, based on reviewer feedback

tylerjereddy commented:
@rsbowen I've pushed in some revisions based on your comments--can you take another look?

If the GitHub GUI allows it, perhaps you can mark resolved review comments as "resolved." If not, maybe let me know which points you think are still outstanding here.

tylerjereddy commented:
As noted at https://github.com/scipy/scipy/pull/20035#discussion_r2047776712, I think there are no remaining disagreements between Richard and I in the actual code here. No old tests have been changed, only new tests added, and this has been open for more than a year without anyone expressing concerns.

The only CI failure is a timeout fluke, and the rebased version of this branch still passes the full suite locally on x86_64 Linux. 

I'll go ahead and self-merge to give this a shot at making the `1.15.3` release. A bit of release manager discretion, but anyway it is me who has to yank/do another release if I mess up.
